### PR TITLE
May Dashboard Cleanup

### PR DESCRIPTION
**Pull request by Abby-Wheelis**
Beginning this PR to address bugs and cleanup in the public dashboard, see list starting here: [#1051](https://github.com/e-mission/e-mission-docs/issues/1051#issuecomment-2106332969)

With respect to the error handling with the survey responses, I believe this is fine, it's been working in the notebooks (and on production, noted since `dfc-fermata` has no data at the moment), but I will keep an eye on it. I am seeing the behavior that only one chart appears in the notebook, but all of the calls to `generate_missing` are going through, and I can see the figures saved in `plots`. 

@shankari re:
>for likert questions, use the related text ("neutral" or "strongly agree") instead of "3" or "5"

We currently have: **"1, 2, 3, 4, 5"**, if we used the related text, it would be **"disagree, -, neutral, -, agree"** due to the formatting of the surveys to appear nicely in the app. I could check for the dashes and give us **"disagree, 2, neutral, 4, agree"** - is that any better? I think it could work if we make sure it is order that way in the legend at least - haven't planned how that would work yet but I'm sure it's possible.

### CODE REVIEW

**Comment by Abby-Wheelis** [comment]
>I think it could work if we make sure it is order that way in the legend at least - haven't planned how that would work yet but I'm sure it's possible.

Trying this, and I am able to order them in the dataframe, but when I pass it into the `plot_and_text_stacked_bar_chart` function, they get ordered again... can I take this ordering out? 

-- commented out for now, how does this look? If there is just a subset of these labels, that is what is shown as well
image

**Comment by shankari** [comment]
@Abby-Wheelis I would suggest the following:
- include the numeric value for all the labels to make it clear (e.g. `Disagree (1)`, and `Neutral(3)`)
- You can pass in a command line argument on whether to sort or not. If you feel comfortable, I would actually suggest moving the sorting to the pre-processing - e.g. `df.groupby('...').agg([...]).apply(calc_pct).sort_values(...)`. The other metrics would then include the sorting step, and these would not.

**Comment by Abby-Wheelis** [comment]
>include the numeric value for all the labels to make it clear (e.g. Disagree (1), and Neutral(3))

Done!

>You can pass in a command line argument on whether to sort or not. If you feel comfortable, I would actually suggest moving the sorting to the pre-processing

I can absolutely do this, but for my own understanding, why do we need to order the metrics in the other charts? The reason that comes to front of mind would be so that they are in the same order between charts?

**Comment by shankari** [comment]
> I can absolutely do this, but for my own understanding, why do we need to order the metrics in the other charts? The reason that comes to front of mind would be so that they are in the same order between charts?

I think it is just to draw the eye towards the biggest bar; and have the ordering be in a logical order instead of arbitrary.
You can try removing it for the other metrics and see what they look like.

Having said all that, I think that we may want to change that sorting anyway when we unify the colors and have them be mapped to the base modes. I will anticipate that we will not want to sort at that point since we will want to have the same ordering for all the charts, which would ideally be based on similar base modes. That's why I am ok with the short-term solution of just passing in an additional argument to the plot function.

**Comment by Abby-Wheelis** [comment]
Piloting a few of the options we've been talking about:

Ordered:
sorted

Unordered:
not_sorted

By count with high participation:
plot_by_count

By count with low participation:
low_labeling

**Comment by Abby-Wheelis** [comment]
side-by-side html tables with a small css change, could increase the padding if wanted, but not by much to avoid cuttoff

we cold also make sure that scroll left-right is available, which will be important when we add inferred (3rd table)

Screenshot 2024-05-29 at 4 55 54 PM

**Comment by Abby-Wheelis** [comment]
What it would look like if all of the survey charts were together:
image

Long survey:
image

I think this would look fine if we split the survey question names into two or three lines ... worried about the longer keys since it's more difficult to put them below now.

**Comment by shankari** [comment]
For the `TripConfirmSurvey`, you should also have the question on the side of the bar. Right now, you have one question at the top (which is really only the question for the last bar) and "Responses" on the side of all the bars.

I do think that the `DfcEvReturnTrip` looks much nicer this way. It gives a sense of the set of questions asked for this survey. And it looks like both the `ReturnTrip` and the `RoamingTrip` have the question "You are satisfied with your choice to take an EV on this trip" (Return shows up here, Roaming shows up on the public dashboard), so grouping the questions by the survey would help clarify that as well.

So assuming you can split the questions into multiple lines, I would suggest that we do group by survey. We should also make sure that the questions for the "Insufficent data" bars are also shown.

**Comment by Abby-Wheelis** [comment]
>For the TripConfirmSurvey, you should also have the question on the side of the bar. Right now, you have one question at the top (which is really only the question for the last bar) and "Responses" on the side of all the bars.

Will do! Sorry, this was one of the first drafts then I kept making changes before I took the 2nd screenshot, I should have clarified that in the comment!

line split the questions:
image

question shown even if no data:
image


Remaining tasks:

- [ ] debug the custom labels that are broken on deployment - bad data?
- [ ] add the new labeling rate visualization
- [ ] test on multiple datasets to look for bugs and:
  - [ ] make sure the labels of "users" vs "tester and participants" are consistent across
  - [ ] make sure the legends work and look nice on all the stacked bar charts
  - [ ] test with sensed but no labeled data to make sure I patched everywhere

**Comment by Abby-Wheelis** [comment]
The issue with `caebike` was the missing `Other` label -- laos has `Other` as a mode, caebike has `Other mode` which is why only some were broken, I have added the mode manually as a part of the label generation process

**Comment by Abby-Wheelis** [comment]
Adding a new plot to show the difference between labeled and sensed over time - 
caebike
image
cortezebikes
image

**Comment by Abby-Wheelis** [comment]
Testing Done:

<details><summary> ca-ebikes (custom labels) </summary>

ca-ebike-default
ca-ebike-stacked
ca-ebike-time
ca-ebike-ebike
ca-ebike-ebike-time

</details>

<details><summary> dfc-fermata (low data, pushes limits of survey space) </summary>

Screenshot 2024-05-30 at 4 33 10 PM
^ I wish these could be dynamically sized, and need to figure out why the axis label is pushed so far below the axis itself

</details>

<details><summary> washingtoncommons (more "normal" survey data, also to test command line since dfc has show tests set to false now) </summary>

Screenshot 2024-05-30 at 4 45 04 PM

</details>

TODO:

- [ ] test with staging or another high-variety dataset to make sure the legends are working
- [ ] clean up the label on the x axis for survey charts (very big lower offset)

@shankari Hoping to have this fully ready for review by tonight, getting down to the last details!

**Comment by Abby-Wheelis** [comment]
I think this works with the legend always by the side - makes sure the mapping is easily visible with each of the bars
image

The way we adjust the positioning of the x axis label is with:
`fig.supxlabel('Proportion (Count)', fontsize=20, x=0.5, y= ax.xaxis.get_label().get_position()[0] - .62, va='top')`

If I lower this number, I can get it to fit right with all of the bars in the surveys - but for just 1 or 2 bars it overlaps the graph. Is there a better way to position it?

Added the labeled vs sensed time series to the survey deployments!

**Comment by Abby-Wheelis** [comment]
Last few changes fixing the html side-by-side and an extra python version change:

Screenshot 2024-05-31 at 9 44 05 AM

This is ready for a review!

**Comment by shankari** [comment]
>  fig.supxlabel('Proportion (Count)', fontsize=20, x=0.5, y= ax.xaxis.get_label().get_position()[0] - .62, va='top')

In general, I am not a big fan of tweaking X and Y positions because, as you found out, it can be complicated to deal with overlaps. Is there a reason that we have to do this instead of just setting the x label? Since we use sharex and sharey, only one x label will be displayed. We don't need a special `supxlabel` call.

**Comment by Abby-Wheelis** [comment]
> Is there a reason that we have to do this instead of just setting the x label? Since we use sharex and sharey, only one x label will be displayed. We don't need a special supxlabel call.

I was not aware of a specific reason - a simple `set_xlabel` works much better!

File: viz_scripts/scaffolding.py
Comment (shankari) [inline viz_scripts/scaffolding.py:204]: Nit, future fix: shouldn't we add the "Other", similar to https://github.com/Abby-Wheelis/em-public-dashboard/commit/75ff38672fa5ed9c16c76a09796faeb9f0d39c31 for mode and replaced mode as well. We don't currently have any deployments with custom values that don't include "Other", but we might!

File: viz_scripts/survey_responses.ipynb
Comment (shankari) [inline viz_scripts/survey_responses.ipynb:348]: not quite sure what this does - are we doing

it looks like we are iterating through the survey names, but then isn't `col` just the question name? A more descriptive variable name would be helpful! Also I think that this can be simplified as

```suggestion
    "        for question_name in quest_dict[survey_name].keys():\n",
```

**Review by shankari — APPROVED** [review]
LGTM! I have a few cleanup suggestions, but nothing big enough to hold up the merge!

File: viz_scripts/scaffolding.py
Comment (Abby-Wheelis) [reply viz_scripts/scaffolding.py:204]: We definitely should make sure each list has "Other" to prevent issues

Comment (Abby-Wheelis) [reply viz_scripts/scaffolding.py:204]: Ah, Other is already added to the modes when we combine replaced and mode -- `combined_mode_values = mode_values + replaced_mode_values + ['Other']` and `combined_mode_values` is what is used to make the color mapping

File: viz_scripts/survey_responses.ipynb
Comment (Abby-Wheelis) [reply viz_scripts/survey_responses.ipynb:348]: I use the `i` to control which position each bar takes - to add all the questions from a given survey to the same chart

I've updated the variable `n` to a more descriptive name `n_questions` and reused it to set the number of inner loops needed

I updated `col` to `quest_name` to be more descriptive

**Pull request merged**
